### PR TITLE
[security] Add ability to crypt data stored to database

### PR DESCRIPTION
**Pull request by makasim**
This is an alternative approach to this PR https://github.com/Payum/Payum/pull/635

Here's some benefits with my approach:

* The algo is not limited to openssl. One can use another one by implementing Cypher interface.
* Rely on `defuse/php-encryption` library. 
* Not only gateway configs could be secured but payment data or any other model (model has to implement CryptedInterface)
* No need for doctrine listener and manipulations with change sets. 
* It would work with previously not encrypted data, it will be encrypted while using.

### CODE REVIEW

**Comment by makasim** [comment]
ping @pjedrzejewski @Arminek @pamil @inssein

**Comment by Arminek** [comment]
Definitely, it is better approach. I felt that this encrypting/decrypting in storage was some kind of workaround.

File: src/Payum/Core/Security/CypherInterface.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+     * @return string
+     */
+    public function decrypt($value);
+
+    /**
+     * This method encrypts the passed value.
+     *
+     * Binary data may be base64-encoded.
+     *
+     * @param string $value
```
Comment (pamil) [inline src/Payum/Core/Security/CypherInterface.php]: Missing return docblock?

File: src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    {
+        return $this->getMock(CypherInterface::class);
+    }
+}
+
+class CryptedModel implements CryptedInterface
+{
+    public function decrypt(CypherInterface $cypher)
+    {
+        $cypher->decrypt('theEncryptedVal');
```
Comment (pamil) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: Missing return

File: src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encrypt(CypherInterface $cypher)
+    {
+        $cypher->encrypt('theVal');
```
Comment (pamil) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: Missing return

File: src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        ;
+        $cypherMock
+            ->expects($this->exactly(2))
+            ->method('decrypt')
+            ->with('theEncryptedVal');
+        ;
+
+        $storage = new CryptoStorageDecorator($decoratedStorage, $cypherMock);
+
+        $storage->findBy([]);
```
Comment (pamil) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: Shouldn't we assert that this method still returns a list of models?

File: src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        ;
+        $cypherMock
+            ->expects($this->once())
+            ->method('decrypt')
+            ->with('theEncryptedVal');
+        ;
+
+        $storage = new CryptoStorageDecorator($decoratedStorage, $cypherMock);
+
+        $storage->find('anId');
```
Comment (pamil) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: Shouldn't we assert that this method still returns a model?

File: src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php (hunk 2 above)
Comment (makasim) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: If you are talking about Crypted::decrypt and Crypted::encrypt methods. They do not return anything by design, just update an internal state of the model.

Comment (pamil) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: Oh, all right, `CypherInterface` and `CryptedInterface` have methods name the same and I just messed it up.

Comment (makasim) [inline src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php]: I felt that it could be that way. The names are easy to mess up.

**Comment by pjedrzejewski** [comment]
Looks great! Thanks for working on this Maksim!

File: src/Payum/Core/Storage/CryptoStorageDecorator.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        return $this->decoratedStorage->identify($model);
+    }
+
+    /**
+     * @param object $model
+     */
+    private function assertCrypted($model)
+    {
+        if (false == $model instanceof  CryptedInterface) {
```
Comment (makasim) [inline src/Payum/Core/Storage/CryptoStorageDecorator.php]: I personally prefer `false ==`. If someone else makes a PR I ignore such things (as it is allowed by PSR1\2). When I code myself I use `false ==`.

**Comment by makasim** [comment]
Merged. Will be available since 1.4 release.

**Pull request merged**